File: tests/styles.py
#!/usr/bin/python

# Copyright (C) 2012 Reece H. Dunn
#
# This file is part of cainteoir-engine.
#
# cainteoir-engine is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# cainteoir-engine is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with cainteoir-engine.  If not, see <http://www.gnu.org/licenses/>.

import harness
import sys

if __name__ == '__main__':
	test = harness.TestSuite('styles', sys.argv)
	test.run({'name': 'CSS', 'groups': [
		{'name': 'Counter Styles', 'type': 'styles', 'tests': [
			{'test': 'css/counter-styles/empty.css', 'result': 'css/counter-styles/empty.styles'},
			{'test': 'css/counter-styles/empty-cased.css', 'result': 'css/counter-styles/empty.styles'},
			{'test': 'css/counter-styles/empty-dual.css', 'result': 'css/counter-styles/empty-dual.styles'},
			{'test': 'css/counter-styles/comments.css', 'result': 'css/counter-styles/empty.styles'},
			{'test': 'css/counter-styles/system-cyclic.css', 'result': 'css/counter-styles/system-cyclic.styles'},
			{'test': 'css/counter-styles/system-cyclic-cased.css', 'result': 'css/counter-styles/system-cyclic.styles'},
			{'test': 'css/counter-styles/system-fixed.css', 'result': 'css/counter-styles/system-fixed.styles'},
			{'test': 'css/counter-styles/system-fixed-cased.css', 'result': 'css/counter-styles/system-fixed.styles'},
			{'test': 'css/counter-styles/system-fixed-offset.css', 'result': 'css/counter-styles/system-fixed-offset.styles'},
			{'test': 'css/counter-styles/system-fixed-negative-offset.css', 'result': 'css/counter-styles/system-fixed-negative-offset.styles'},
			{'test': 'css/counter-styles/system-symbolic.css', 'result': 'css/counter-styles/empty.styles'},
			{'test': 'css/counter-styles/system-symbolic-cased.css', 'result': 'css/counter-styles/empty.styles'},
			{'test': 'css/counter-styles/system-alphabetic.css', 'result': 'css/counter-styles/system-alphabetic.styles'},
			{'test': 'css/counter-styles/system-alphabetic-cased.css', 'result': 'css/counter-styles/system-alphabetic.styles'},
			{'test': 'css/counter-styles/system-numeric.css', 'result': 'css/counter-styles/system-numeric.styles'},
			{'test': 'css/counter-styles/system-numeric-cased.css', 'result': 'css/counter-styles/system-numeric.styles'},
			{'test': 'css/counter-styles/system-additive.css', 'result': 'css/counter-styles/system-additive.styles'},
			{'test': 'css/counter-styles/system-additive-cased.css', 'result': 'css/counter-styles/system-additive.styles'},
			{'test': 'css/counter-styles/negative.css', 'result': 'css/counter-styles/negative.styles'},
			{'test': 'css/counter-styles/negative-cased.css', 'result': 'css/counter-styles/negative.styles'},
			{'test': 'css/counter-styles/negative-empty.css', 'result': 'css/counter-styles/negative-empty.styles'},
			{'test': 'css/counter-styles/prefix.css', 'result': 'css/counter-styles/prefix.styles'},
			{'test': 'css/counter-styles/prefix-cased.css', 'result': 'css/counter-styles/prefix.styles'},
			{'test': 'css/counter-styles/prefix-empty.css', 'result': 'css/counter-styles/prefix-empty.styles'},
			{'test': 'css/counter-styles/suffix.css', 'result': 'css/counter-styles/suffix.styles'},
			{'test': 'css/counter-styles/suffix-cased.css', 'result': 'css/counter-styles/suffix.styles'},
			{'test': 'css/counter-styles/suffix-empty.css', 'result': 'css/counter-styles/suffix-empty.styles'},
			{'test': 'css/counter-styles/symbols.css', 'result': 'css/counter-styles/symbols.styles'},
			{'test': 'css/counter-styles/symbols-cased.css', 'result': 'css/counter-styles/symbols.styles'},
			{'test': 'css/counter-styles/additive-symbols.css', 'result': 'css/counter-styles/additive-symbols.styles'},
			{'test': 'css/counter-styles/additive-symbols-cased.css', 'result': 'css/counter-styles/additive-symbols.styles'},
			{'test': 'css/counter-styles/range.css', 'result': 'css/counter-styles/range.styles'},
			{'test': 'css/counter-styles/range-cased.css', 'result': 'css/counter-styles/range.styles'},
			{'test': 'css/counter-styles/disc.css', 'result': 'css/counter-styles/disc.styles'},
		]},
	]})
	test.summary()

File: tests/styles.cpp
/* Test driver for style manager (CSS) tests.
 *
 * Copyright (C) 2012 Reece H. Dunn
 *
 * This file is part of cainteoir-engine.
 *
 * cainteoir-engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cainteoir-engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cainteoir-engine.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <cainteoir/content.hpp>
#include <stdexcept>

std::string get_counter_style_name(const cainteoir::style_manager &aStyles, const cainteoir::counter_style *aStyle)
{
	for (auto &item : aStyles.counter_styles())
	{
		if (item.second == aStyle)
			return item.first;
	}
	return "decimal";
}

const char *get_counter_system_string(cainteoir::counter_system system)
{
	using cainteoir::counter_system;
	switch (system)
	{
	case counter_system::cyclic:     return "cyclic";
	case counter_system::fixed:      return "fixed";
	case counter_system::symbolic:   return "symbolic";
	case counter_system::alphabetic: return "alphabetic";
	case counter_system::numeric:    return "numeric";
	case counter_system::additive:   return "additive";
	default: return nullptr;
	}
}

int main(int argc, char ** argv)
{
	try
	{
		argc -= 1;
		argv += 1;

		if (argc == 0)
			throw std::runtime_error("no document specified");

		cainteoir::style_manager styles;
		styles.parse(cainteoir::make_file_buffer(argv[0]));
		for (auto &item : styles.counter_styles())
		{
			fprintf(stdout, "@counter-style %s {\n", item.first.c_str());
			auto &style = *item.second;
			if (style.system == cainteoir::counter_system::fixed)
				fprintf(stdout, "\tsystem: fixed %d;\n", style.initial_symbol_value);
			else
				fprintf(stdout, "\tsystem: %s;\n", get_counter_system_string(style.system));
			fprintf(stdout, "\tnegative: '%s';\n", style.negative.c_str());
			fprintf(stdout, "\tprefix: '%s';\n", style.prefix.c_str());
			fprintf(stdout, "\tsuffix: '%s';\n", style.suffix.c_str());
			fprintf(stdout, "\trange:");
			if (style.range == cainteoir::counter_style::get_auto_range(style.system))
				fprintf(stdout, " auto");
			else
			{
				if (style.range.first == cainteoir::counter_style::infinite.first)
					fprintf(stdout, " infinite");
				else
					fprintf(stdout, " %d");
				if (style.range.second == cainteoir::counter_style::infinite.second)
					fprintf(stdout, " infinite");
				else
					fprintf(stdout, " %d");
			}
			fprintf(stdout, ";\n");
			fprintf(stdout, "\tsymbols:");
			for (auto &symbol : style.symbols)
				fprintf(stdout, " '%s'", symbol.c_str());
			fprintf(stdout, " ;\n");
			fprintf(stdout, "\tadditive-symbols:");
			bool need_comma = false;
			for (auto &symbol : style.additive_symbols)
			{
				if (need_comma)
					fprintf(stdout, ",");
				fprintf(stdout, " %d '%s'", symbol.first, symbol.second.c_str());
				need_comma = true;
			}
			fprintf(stdout, " ;\n");
			fprintf(stdout, "\tfallback: %s;\n", get_counter_style_name(styles, style.fallback).c_str());
			fprintf(stdout, "}\n\n");
		}
	}
	catch (std::runtime_error &e)
	{
		fprintf(stderr, "error: %s\n", e.what());
	}

	return 0;
}
